src/ir/module-utils.h

```diff
@@ -318,26 +330,68 @@ template<typename T> struct ParallelFunctionMap {
     }
 
     // Run on the implemented functions.
+    struct MapperInfo {
+      Module* module;
+      Map* map;
+      Func work;
+    };
+
     struct Mapper : public WalkerPass<PostWalker<Mapper>> {
       bool isFunctionParallel() override { return true; }
 
-      Mapper(Info* info) : info(info) {}
+      Mapper(MapperInfo* info) : info(info) {}
 
       Mapper* create() override { return new Mapper(info); }
 
-      void doWalkFunction(Function* curr) {
+      void visitCall(Call* curr) {
+        (*info->map)[this->getFunction()].callsTo.insert(
+          info->module->getFunction(curr->target));
+      }
+
+      void visitFunction(Function* curr) {
         assert((*info->map).count(curr));
         info->work(curr, (*info->map)[curr]);
       }
 
     private:
-      Info* info;
+      MapperInfo* info;
     };
 
-    Info info = {&map, work};
+    MapperInfo info = {&wasm, &map, work};
 
     PassRunner runner(&wasm);
     Mapper(&info).run(&runner, &wasm);
+
+    // Find what is called by what.
+    for (auto& pair : map) {
+      auto* func = pair.first;
+      auto& info = pair.second;
+      for (auto* target : info.callsTo) {
+        map[target].calledBy.insert(func);
+      }
+    }
+  }
+
+  // Propagate a property from a function to those that call it.
+  void propagateBack(std::function<bool(const T&)> hasProperty,
+                     std::function<bool(const T&)> canHaveProperty,
+                     std::function<void(T&)> addProperty) {
+    // The work queue contains an item we just learned can change the state.
+    UniqueDeferredQueue<Function*> work;
+    for (auto& func : wasm.functions) {
+      if (hasProperty(map[func.get()])) {
+        work.push(func.get());
+      }
+    }
+    while (!work.empty()) {
+      auto* func = work.pop();
+      for (auto* caller : map[func].calledBy) {
+        if (!hasProperty(map[caller]) && canHaveProperty(map[caller])) {
```

Comment:
- Given that we are gonna set the property regardless of whether we currently have the property or not, do we need `hasProperty` here?
- Nit: How about `needsProperty` in place of `canHaveProperty`? `canHaveProperty` sounds like it may or may not have the property...

Reply:
We check if we already have the property before setting it. That is, this line checks "do we not already have the property, and can we receive it? If so, then it propagates back to it now." (Maybe I misunderstood your comment?)

Reply:
Oh you're right. What I meant was, I was wondering if we need to check `!hasProperty` here, but we need to, because we don't want to add the same caller twice to the work queue... Sorry, nevermind.